pulsar-proxy/src/test/java/org/apache/pulsar/proxy/server/ProxyWithExtensibleLoadManagerTest.java

```diff
@@ -0,0 +1,229 @@
+
+public class ProxyWithExtensibleLoadManagerTest extends MultiBrokerBaseTest {
+
+    private static final int TEST_TIMEOUT_MS = 30_000;
+
+    private ProxyService proxyService;
+
+    @Override
+    public int numberOfAdditionalBrokers() {
+        return 1;
+    }
+
+    @Override
+    public void doInitConf() throws Exception {
+        super.doInitConf();
+        configureExtensibleLoadManager(conf);
+    }
+
+    @Override
+    protected ServiceConfiguration createConfForAdditionalBroker(int additionalBrokerIndex) {
+        return configureExtensibleLoadManager(getDefaultConf());
+    }
+
+    private ServiceConfiguration configureExtensibleLoadManager(ServiceConfiguration config) {
+        config.setNumIOThreads(8);
+        config.setLoadBalancerInFlightServiceUnitStateWaitingTimeInMillis(5 * 1000);
+        config.setLoadBalancerServiceUnitStateMonitorIntervalInSeconds(1);
+        config.setLoadManagerClassName(ExtensibleLoadManagerImpl.class.getName());
+        config.setLoadBalancerSheddingEnabled(false);
+        return config;
+    }
+
+    private ProxyConfiguration initializeProxyConfig() {
+        var proxyConfig = new ProxyConfiguration();
+        proxyConfig.setNumIOThreads(8);
+        proxyConfig.setServicePort(Optional.of(0));
+        proxyConfig.setBrokerProxyAllowedTargetPorts("*");
+        proxyConfig.setMetadataStoreUrl(DUMMY_VALUE);
+        proxyConfig.setConfigurationMetadataStoreUrl(GLOBAL_DUMMY_VALUE);
+        return proxyConfig;
+    }
+
+    private LookupService spyLookupService(PulsarClient client) throws IllegalAccessException {
+        LookupService svc = (LookupService) FieldUtils.readDeclaredField(client, "lookup", true);
+        var lookup = spy(svc);
+        FieldUtils.writeDeclaredField(client, "lookup", lookup, true);
+        return lookup;
+    }
+
+    private PulsarClientImpl createClient(ProxyService proxyService) {
+        try {
+            return Mockito.spy((PulsarClientImpl) PulsarClient.builder().
+                    serviceUrl(proxyService.getServiceUrl()).
+                    build());
+        } catch (PulsarClientException e) {
+            throw new CompletionException(e);
+        }
+    }
+
+    @BeforeMethod(alwaysRun = true)
+    public void proxySetup() throws Exception {
+        var proxyConfig = initializeProxyConfig();
+        proxyService = Mockito.spy(new ProxyService(proxyConfig, new AuthenticationService(
+                PulsarConfigurationLoader.convertFrom(proxyConfig))));
+        doReturn(registerCloseable(new ZKMetadataStore(mockZooKeeper))).when(proxyService).createLocalMetadataStore();
+        doReturn(registerCloseable(new ZKMetadataStore(mockZooKeeperGlobal))).when(proxyService)
+                .createConfigurationMetadataStore();
+        proxyService.start();
+    }
+
+    @AfterMethod(alwaysRun = true)
+    public void proxyCleanup() throws Exception {
+        if (proxyService != null) {
+            proxyService.close();
+        }
+    }
+
```

Comment:
Could we add a case to assert that the clients can connect to the brokers directly when we shutdown the proxy in the middle,?

Reply:
Let's say a user is upgrading the cluster to migrate from pulsar-proxy to something else(istio). In this case, the connected clients should get connection errors from the proxy (as the proxy is shut down). Then, the clients should try to reconnect and soon get the broker addresses (assuming proxy hosts are removed under the service url and the brokers are added under it).


To simulate this case, I guess we need the following mocks.

0. mock resolveHost() to return the proxy address when called the first time and return the broker address when called the second time.
1. let the client connect to the proxy (assert useProxy is true)
3. shutdown proxy
4. assert that the client connects to the broker (assert useProxy is false)

Reply:
However, for istio -> proxy migration, I don't think this graceful reconnection is possible without broker restarts(we need to forcefully unload all clients to reconnect)